homeassistant/components/pi_hole/config_flow.py

```diff
@@ -136,22 +97,38 @@ async def async_step_init(
             errors=errors,
         )
 
-    async def async_step_api_key(
-        self, user_input: dict[str, Any] | None = None
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self._config = self.init_data
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self,
+        user_input: dict[str, Any] | None = None,
     ) -> FlowResult:
-        """Handle step to setup API key."""
+        """Perform reauth confirm upon an API authentication error."""
+        errors = {}
         if user_input is not None:
-            return self.async_create_entry(
-                title=self._config[CONF_NAME],
-                data={
-                    **self._config,
-                    CONF_API_KEY: user_input.get(CONF_API_KEY, ""),
-                },
-            )
+            self._config = {**self._config, CONF_API_KEY: user_input[CONF_API_KEY]}
+            if not (errors := await self._async_try_connect()):
+                entry = self.hass.config_entries.async_get_entry(
+                    self.context["entry_id"]
+                )
+                assert entry
+                self.hass.config_entries.async_update_entry(entry, data=self._config)
+                self.hass.async_create_task(
+                    self.hass.config_entries.async_reload(self.context["entry_id"])
+                )
+                return self.async_abort(reason="reauth_successful")
 
         return self.async_show_form(
-            step_id="api_key",
-            data_schema=vol.Schema({vol.Optional(CONF_API_KEY): str}),
+            step_id="reauth_confirm",
+            description_placeholders={
+                CONF_HOST: self._config[CONF_HOST],
+                CONF_LOCATION: self._config[CONF_LOCATION],
+            },
+            data_schema=vol.Schema({vol.Required(CONF_API_KEY): str}),
+            errors=errors,
         )
 
     async def _async_endpoint_existed(self, endpoint: str) -> bool:
```

Comment:
Side note: We can probably replace this method by using this helper:
https://github.com/home-assistant/core/blob/77dd0b2fd3da885571f4d78f4ab23b6531d2b3c3/homeassistant/config_entries.py#L1376-L1380

Reply:
nice, thanks for this hint -> [`8703439`](https://github.com/home-assistant/core/pull/85264/commits/87034390b400d639cdc3406022452a2e8db1ada4)